File: 1. Spark Familiarization/submission/task3.py
import sys
import json
import re
from pyspark import SparkContext
sc = SparkContext.getOrCreate()

if __name__ == '__main__':
  review = sc.textFile(sys.argv[1])
  output = sys.argv[2]
  partition_type = sys.argv[3]
  n_partition = int(sys.argv[4])
  n = int(sys.argv[5])
  ans = dict()
  if partition_type == 'default':
    ans['n_partitions'] = review.getNumPartitions()
    ans['n_items'] = review.glom().map(lambda x:len(x)).collect()
    ans['result'] = review.map(lambda x:(json.loads(x)['business_id'],1)).reduceByKey(lambda a,b:a+b).filter(lambda x:x[1]>n).collect()
  else:
    '''
    def reduce(iterators):
        c = {}
        for iter in iterators:
            if iter[0] not in c:
                c[iter[0]]=1
            else:
                c[iter[0]]+=1
        for i in c:
            yield (i,c[i])
    '''
    ans['n_partitions'] = n_partition
    a = review.map(lambda x:(json.loads(x)['business_id'],1)).partitionBy(n_partition).persist()
    ans['n_items'] = a.glom().map(len).collect()
    ans['result'] = a.reduceByKey(lambda x,y:x+y).filter(lambda x:x[1]>n).collect()
    
  with open(output,"w") as file:
    file.write(json.dumps(ans))

File: 1. Spark Familiarization/submission/task3.scala
import org.json4s.jackson.JsonMethods._
import org.json4s._
import org.json4s.jackson.Serialization.write
import org.json4s.JsonDSL._
import org.apache.spark.sql.SparkSession
import java.io._
import scala.collection.mutable.ListBuffer
import org.apache.spark.HashPartitioner

object task3 {
  def main(args: Array[String]): Unit = {
    //val t0 = System.nanoTime()
    implicit val formats = org.json4s.DefaultFormats
    val ss = SparkSession
      .builder()
      .appName("scala")
      .config("spark.master", "local[*]")
      .getOrCreate()
    val sc = ss.sparkContext
    val review = sc.textFile(args(0))
    val partition_type = args(2)
    val n_partition = args(3).toInt
    val n = args(4).toInt
    var ans = JObject()
    if (partition_type == "default") {
      ans = ans ~ ("n_partitions", review.getNumPartitions)
      ans = ans ~ ("n_items", review.glom().map(x => x.length).collect().toList)
      val a = review.map(x => ((parse(x) \\ "business_id").values.toString, 1)).reduceByKey(_ + _).filter(_._2 > n).collect()
      val k = a.map(t => List(JString(t._1), JInt(t._2))).toList
      ans = ans ~ ("result", k)
    }else{
      ans = ans ~ ("n_partitions", n_partition)
      val cus = review.map(x => (parse(x) \\ "business_id").values.toString).map(x => (x,1)).partitionBy(new HashPartitioner(n_partition)).persist()
      ans = ans ~ ("n_items",cus.glom().map(_.length).collect().toList)
      val a = cus.reduceByKey(_ + _).filter(_._2 > n).collect()
      val k = a.map(t => List(JString(t._1), JInt(t._2))).toList
      ans = ans ~ ("result", k)
      
      
    }
    val file = new File(args(1))
    val output = new BufferedWriter(new FileWriter(file))
    output.write(write(ans))
    output.close()
    //val t1 = System.nanoTime()
    //println("Duration: ",(t1-t0)/10^6)
  }
}

